rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -136,42 +156,268 @@ def consume_result(self, used_amount: FVal) -> None:
+                taxable=taxable,
+            ))
+            used_acquisitions.append(acquisition_event)
+            self.consume_result(acquisition_event.remaining_amount)
+            # and since this event is going to be removed, reduce its remaining to zero
+            acquisition_event.remaining_amount = ZERO
+
+        is_complete = True
+        if remaining_sold_amount != ZERO:
+            # if we still have sold amount but no acquisitions to satisfy it then we only
+            # found acquisitions to partially satisfy the sell
+            adjusted_amount = spending_amount - taxfree_amount
+            missing_acquisitions.append(
+                MissingAcquisition(
+                    asset=spending_asset,
+                    time=timestamp,
+                    found_amount=taxable_amount + taxfree_amount,
+                    missing_amount=remaining_sold_amount,
+                ),
+            )
+            taxable_amount = adjusted_amount
+            is_complete = False
+
+        return CostBasisInfo(
+            taxable_amount=taxable_amount,
+            taxable_bought_cost=taxable_bought_cost,
+            taxfree_bought_cost=taxfree_bought_cost,
+            matched_acquisitions=matched_acquisitions,
+            is_complete=is_complete,
+        )
 
     def __len__(self) -> int:
         return len(self._acquisitions)
 
 
-class FIFOAcquisitionsOrder(BaseAcquisitionsOrder):
-    """Accounting in FIFO (first-in-first-out) order"""
+class FIFOCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in FIFO (first-in-first-out) method"""
+    _count: int
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = 0
+
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.append(acquisition)
+        heapq.heappush(self._acquisitions, (self._count, acquisition))
+        self._count += 1
+
 
+class LIFOCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in LIFO (last-in-first-out) method"""
+    _count: int
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = 0
 
-class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
-    """Accounting in LIFO (last-in-first-out) order"""
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.appendleft(acquisition)
+        heapq.heappush(self._acquisitions, (-self._count, acquisition))
+        self._count += 1
+
+
+class HIFOCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in HIFO (highest-in-first-out) method"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        heapq.heappush(self._acquisitions, (-acquisition.amount, acquisition))
+
+
+class AverageCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in Average Cost Base(ACB) method"""
+    # the average cost base of last event(buy/sell) that occurred
+    current_average_cost_base: FVal
```

Comment:
Here too ... and generally search and replace. Won't point all of them

```suggestion
    current_average_cost_basis: FVal
```